### PR TITLE
ConvLayer, support 1d conv with striding and dilation

### PR DESCRIPTION
**Pull request by dierkes-j**
Implemented 1d convolution with striding and dilation

@vieting

### CODE REVIEW

File: tests/test_TFNetworkLayer.py [hunk 1]
```diff
@@ -6806,6 +6806,39 @@ def test_ConvLayer_empty_out():
...
+    conv_weight = torch.ones(7, 5, 2)
+    torch_out = torch.nn.functional.conv1d(
+      torch_input.transpose(1,2), conv_weight, bias=None, stride=2, padding=0, dilation=2
+    ).transpose(1,2).numpy()
+    print(out.shape)
+    print(seq_lens)
+    assert isinstance(out, numpy.ndarray)
+    assert isinstance(seq_lens, numpy.ndarray)
+    assert (torch_out == out).all()
+
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:6841]: Missing newline here (2 newlines between functions).

File: returnn/tf/layers/basic.py [hunk 2]
```diff
@@ -5218,10 +5218,27 @@ def __init__(self, filter_size, padding, strides=1, dilation_rate=1, groups=1,
         strides = strides[:-1]
         dilation_rate = dilation_rate[:-1]
     else:
-      y = tf_compat.v1.nn.convolution(
-        x, data_format=data_format,
-        filter=filters,
-        padding=padding, strides=strides, dilation_rate=dilation_rate)
+      if max(strides) > 1 and max(dilation_rate) > 1:
```
Comment (albertz) [inline returnn/tf/layers/basic.py:5221]: Make this better an `elif ...` and merge with the line above.

File: returnn/tf/layers/basic.py [hunk 3]
```diff
@@ -5218,10 +5218,27 @@ def __init__(self, filter_size, padding, strides=1, dilation_rate=1, groups=1,
...
+        # tf.nn.convolution does not support this, therefore resort to tf.nn.conv2d which requires some adaptations
+        from_conv1d = False
```
Comment (albertz) [inline returnn/tf/layers/basic.py:5223]:
```suggestion
        squeeze_axis = None
```

File: returnn/tf/layers/basic.py [hunk 4]
```diff
@@ -5218,10 +5218,27 @@ def __init__(self, filter_size, padding, strides=1, dilation_rate=1, groups=1,
...
+            x = tf.expand_dims(x, axis=-2)
+            squeeze_axis = -2
+          else:   # we are in the case NWC
+            data_format = "NHWC"
+            filters = tf.expand_dims(filters, axis=0)
+            x = tf.expand_dims(x, axis=-3) # data format is NHWC
+            squeeze_axis = -3
+        assert data_format in ["NCHW", "NHWC"], "not implemented otherwise"
+        y = tf_compat.v1.nn.conv2d(x, data_format=data_format, filter=filters, padding=padding, strides=strides, dilations=dilation_rate)
+        if from_conv1d:
```
Comment (albertz) [inline returnn/tf/layers/basic.py:5238]:
```suggestion
        if squeeze_axis is not None:
```

**Comment by albertz** [comment]
Please check the code style. Do not ignore the warnings by your editor.
```
  returnn/tf/layers/basic.py:5234: WEAK WARNING Pep8CodeStyle: E261: at least two spaces before inline comment
  returnn/tf/layers/basic.py:5237: WEAK WARNING Pep8CodeStyle: E501: line too long (137 > 120 characters)
  returnn/tf/layers/basic.py:5239: WARNING PyUnboundLocalVariableInspection: Local variable 'squeeze_axis' might be referenced before assignment
  returnn/tf/layers/basic.py:5241: WEAK WARNING Pep8CodeStyle: E501: line too long (146 > 120 characters)
```

**Comment by albertz** [comment]
What was the error without this PR? Can you post this here?

File: tests/test_TFNetworkLayer.py [hunk 5]
```diff
@@ -6806,6 +6806,39 @@ def test_ConvLayer_empty_out():
     assert out.shape == (1, 0, 7)
 
 
+def test_ConvLayer_stride_dilation():
+  import numpy as np
+  import torch
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:6811]: Do not use PyTorch in the test.

**Comment by dierkes-j** [comment]
> What was the error without this PR? Can you post this here?

The error was thrown by TensorFlow for a 1d convolution: `strides > 1 not supported in conjunction with dilation_rate > 1`

File: tests/test_TFNetworkLayer.py (hunk 5 above)
Comment (dierkes-j) [reply tests/test_TFNetworkLayer.py:6811]: Whats best then: Should I use a different library than PyTorch/TensorFlow or should I simply hardcode the expected result of the convolution?

Comment (albertz) [reply tests/test_TFNetworkLayer.py:6811]: You can implement it manually using NumPy. Or you can hardcode the expected result. Whatever is easier. The numbers should be as minimal as possible anyway.

Comment (albertz) [reply tests/test_TFNetworkLayer.py:6811]: Or you can just ignore this also. It's not really so important.

**Comment by albertz** [comment]
Please check that your code follows proper code style.
```
  returnn/tf/layers/basic.py:5244: WEAK WARNING Pep8CodeStyle: W291: trailing whitespace
  returnn/tf/layers/basic.py:5249: WEAK WARNING Pep8CodeStyle: W291: trailing whitespace
```

File: tests/test_TFNetworkLayer.py [hunk 6]
```diff
@@ -6806,6 +6806,35 @@ def test_ConvLayer_empty_out():
     assert out.shape == (1, 0, 7)
 
 
+def test_ConvLayer_stride_dilation():
+  import numpy as np
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:6810]: You don't need this import. We already import `numpy`.

File: returnn/tf/layers/basic.py [hunk 7]
```diff
@@ -5217,11 +5217,28 @@ def __init__(self, filter_size, padding, strides=1, dilation_rate=1, groups=1,
...
+                                 strides=strides, dilations=dilation_rate)
+      if squeeze_axis is not None:
+        y = tf.squeeze(y, axis=squeeze_axis)
     else:
-      y = tf_compat.v1.nn.convolution(
-        x, data_format=data_format,
-        filter=filters,
-        padding=padding, strides=strides, dilation_rate=dilation_rate)
+      y = tf_compat.v1.nn.convolution(x, data_format=data_format, filter=filters, padding=padding, 
+                                      strides=strides, dilation_rate=dilation_rate)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:5241]: Do not reformat it like this. Do not indent it like this.
```suggestion
      y = tf_compat.v1.nn.convolution(
        x, data_format=data_format, filter=filters, padding=padding, 
        strides=strides, dilation_rate=dilation_rate)
```

**Review by albertz — APPROVED** [review]

**Pull request merged**